test/ha/kourier_control_test.go

```diff
@@ -0,0 +1,88 @@
+// +build e2e
+
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ha
+
+import (
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/ptr"
+	pkgTest "knative.dev/pkg/test"
+	pkgHa "knative.dev/pkg/test/ha"
+	"knative.dev/serving/pkg/apis/networking"
+	"knative.dev/serving/test/conformance/ingress"
+	"knative.dev/serving/test/e2e"
+)
+
+const (
+	ingressNamespace         = "kourier-system"
+	kourierHPALease          = "kourier"
+	kourierControlDeployment = "3scale-kourier-control"
+)
+
+func TestKourierControlHA(t *testing.T) {
+	clients := e2e.Setup(t)
+
+	if err := pkgTest.WaitForDeploymentScale(clients.KubeClient, kourierControlDeployment, ingressNamespace, haReplicas); err != nil {
+		t.Fatalf("Deployment %s not scaled to %d: %v", kourierControlDeployment, haReplicas, err)
+	}
+
+	leaderController, err := pkgHa.GetLeader(clients.KubeClient, kourierHPALease, ingressNamespace)
+	if err != nil {
+		t.Fatalf("Failed to get leader: %v", err)
+	}
+
+	svcName, svcPort, svcCancel := ingress.CreateRuntimeService(t, clients, networking.ServicePortNameHTTP1)
+	defer svcCancel()
+
+	_, _, ingressCancel := ingress.CreateIngressReady(t, clients, createIngressSpec(svcName, svcPort))
+	defer ingressCancel()
+
+	url := apis.HTTP(svcName + domain)
+
+	assertIngressEventuallyWorks(t, clients, url.URL())
+
+	if err := clients.KubeClient.Kube.CoreV1().Pods(ingressNamespace).Delete(leaderController, &metav1.DeleteOptions{
+		GracePeriodSeconds: ptr.Int64(0),
+	}); err != nil {
+		t.Fatalf("Failed to delete pod %s: %v", leaderController, err)
+	}
+
+	if err := pkgTest.WaitForPodDeleted(clients.KubeClient, leaderController, ingressNamespace); err != nil {
+		t.Fatalf("Did not observe %s to actually be deleted: %v", leaderController, err)
+	}
+
+	// Make sure a new leader has been elected.
+	if _, err = pkgHa.GetLeader(clients.KubeClient, kourierHPALease, ingressNamespace); err != nil {
+		t.Fatalf("Failed to find new leader: %v", err)
+	}
+
+	assertIngressEventuallyWorks(t, clients, url.URL())
```

Comment:
See my comment [here](https://github.com/knative/serving/pull/7325/files#r420846913), but these checks aren't meaningfully checking anything, and I worry about their proliferation.

The check below is a bit better (it at least checks that the control plane comes back), but it still isn't actually testing leader election.

If it weren't for the lease lookups (which only test initial creation), I believe this test should pass without leader election enabled 😬

Do you want to spend some cycles chatting about ways to effectively test this?

Reply:
As discussed on slack, I've renamed the GetLeader function to WaitForNewLeader and added one more parameter so that we can compare with the previous leader. The logic is same as before but it's more clear now that we're checking that there's a new holder of the lease.

Reply:
The other thing I mentioned in slack was running a dataplane probe during the test.

This relates to a concern I was also bugging @markusthoemmes about (maybe @davidor knows?): are there are any safeguards in Kourier to keep it from serving xDS to Envoy before it has read all of the KIngress configuration into its cache?

If it does, then my concern is that a new leader could wipe out the network programming for your Envoys during a failover, which is probably worse than the Envoys not being able to get new programming.

Reply:
Alright. Pushed a commit with the prober test. I'm not sure if there won't be any failed requests. Let's see.